### PR TITLE
Add tabletop computers

### PR DESCRIPTION
**Pull request by I-JustUser-I**
## About the PR

This PR adds tabletop versions computers. Inspired by tabletop computers from Frontier.

## Why / Balance

This will help mappers and players create combinations of a wide variety of tables with tabletop computers. These computers will make the bridge and shuttles look more interesting!
## Technical details

I conducted several playtests this change with other players and found no errors, this PR is completely correct. A new component has been added to the computer board, similar to a component on a regular computer. The code is optimized using the main computer prototype as a parent. The resprite was made by me based on current computer sprites. Added check for tabletop computer in BoardNodeEntity.cs.

## Media

AllTabletopComputers
Crafting
EntitySpawnPanel

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase
**Changelog**

:cl: JustUser
- add: Added tabletop computers!

### CODE REVIEW

**Comment by UbaserB** [comment]
I’ve been wanting these since I saw them in frontier, blessed.

File: Content.Server/Construction/NodeEntities/BoardNodeEntity.cs [hunk 1]
```diff
@@ -29,6 +29,14 @@ public sealed partial class BoardNodeEntity : IGraphNodeEntity
 
         var board = container.ContainedEntities[0];
 
+        //tabletop computer check
+        if (args.EntityManager.TryGetComponent(container.Owner, out ConstructionComponent? constructionComponent)
+            && constructionComponent.Graph == "TabletopComputer"
```
Comment (deathride58) [inline Content.Server/Construction/NodeEntities/BoardNodeEntity.cs:34]: Please do not hardcode specific prototype strings into systems like this.

File: Content.Server/Construction/Components/TabletopComputerBoardComponent.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom.Prototype;
+
+namespace Content.Server.Construction.Components
+{
+    [RegisterComponent, ComponentProtoName("TabletopComputerBoardComponent")]
+    public sealed partial class TabletopComputerBoardComponent : Component
```
Comment (deathride58) [inline Content.Server/Construction/Components/TabletopComputerBoardComponent.cs:7]: Is there any reason why TabletopComputerBoard needs to be it's own bespoke component? This leads to a *massive* amount of copy-paste in construction graph prototypes.

File: Resources/Textures/Structures/Machines/tabletop_computers.rsi/meta.json [hunk 3]
```diff
@@ -0,0 +1,27 @@
+{
+  "version": 1,
+  "license": "CC-BY-SA-3.0",
+  "copyright": "Taken from tgstation at commit https://github.com/tgstation/tgstation/commit/bd6873fd4dd6a61d7e46f1d75cd4d90f64c40894, computers modified by JustUser.",
```
Comment (deathride58) [inline Resources/Textures/Structures/Machines/tabletop_computers.rsi/meta.json:4]: Uh, no. These are not your sprites.
https://github.com/new-frontiers-14/frontier-station-14/blob/master/Resources/Textures/_NF/Structures/Machines/computer_tabletop.rsi/meta.json#L4

Please do not misattribute CC-BY-SA work.
```suggestion
  "copyright": "Taken from tgstation at commit https://github.com/tgstation/tgstation/commit/bd6873fd4dd6a61d7e46f1d75cd4d90f64c40894. | Resprited by erhardsteinhauer (discord)",
```

File: Resources/Textures/Structures/Machines/tabletop_parts.rsi/meta.json [hunk 4]
```diff
@@ -0,0 +1,31 @@
+{
+  "version": 1,
+  "size": {
+    "x": 32,
+    "y": 32
+  },
+  "license": "CC-BY-SA-3.0",
+  "copyright": "Taken from tgstation at https://github.com/tgstation/tgstation/commit/0d9c9a8233dfc3fc55edc538955a761a6328bee0, computer frames modified by JustUser",
```
Comment (deathride58) [inline Resources/Textures/Structures/Machines/tabletop_parts.rsi/meta.json:8]: These are not your sprites. Please attribute the original authors properly as per the license terms of the CC-BY-SA.
```suggestion
  "copyright": "Taken from tgstation at commit https://github.com/tgstation/tgstation/commit/bd6873fd4dd6a61d7e46f1d75cd4d90f64c40894. | Resprited by erhardsteinhauer (discord)",
```

**Review by deathride58 — CHANGES_REQUESTED** [review]
On top of this PR being an unattributed copy-paste of Frontier's code with names shifted around in an attempt to throw off attempts to find the source, this PR contains violations of the CC-BY-SA license. Please fix this.

Additionally, it feels like there's *much* better ways to go about implementing the concept of tabletop computers. It's not explicitly required to do so, but it'd be far cleaner to opt to have the tabletop sprite variants automatically swap in if a computer is present on the same tile as a table; that'd *vastly* reduce the amount of copy-pasting prevalent in this PR, and would greatly improve the maintainability of the feature.

File: Content.Server/Construction/Components/TabletopComputerBoardComponent.cs (hunk 2 above)
Comment (I-JustUser-I) [reply Content.Server/Construction/Components/TabletopComputerBoardComponent.cs:7]: I tried adding this to an existing ComputerComponent, but I couldn't get it to work, so I had to use this option. I agree that this should have been avoided, but I couldn't...

**Comment by I-JustUser-I** [comment]
Thanks for the review and I agree with a lot of it. I don’t deny that I looked at the frontier implementation, but it was too cluttered and I tried to make my own clean one, which seems to have worked out. Copying the code from there is minimal, I only copied the check in craft, since I could not make the implementation better. Almost all the code is a version of the original code with strong modifications, but the implementation is really very similar. With the license, I didn’t want to appropriate someone else’s work, although I really did the resprites, but the original is definitely not mine and I didn’t know what to do with the license, sorry.

**Comment by I-JustUser-I** [comment]
As for building the main computer and checking whether it is on the table... it seems to me that this is not the most obvious for players, many players are already unaware of a variety of functions, since they are not very obvious, I tried to make sure that the computers were understandable and easy to reproduce.

File: Content.Server/Construction/Components/TabletopComputerBoardComponent.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom.Prototype;
+
+namespace Content.Server.Construction.Components
```
Comment (Simyon264) [inline Content.Server/Construction/Components/TabletopComputerBoardComponent.cs:4]: Use filescoped namespaces

**Review by Simyon264 — CHANGES_REQUESTED** [review]

**Comment by ErhardSteinhauer** [comment]
> I don’t deny that I looked at the frontier

Yet never explicitly state in this PR your sources of inspiration.

**Comment by I-JustUser-I** [comment]
I agree, I'll fix it!

**Comment by keronshb** [comment]
I would rather see actual laptop implementation from TG/Cit SS13 with limited programs, but that's an entire system of it's own.

File: Content.Server/Construction/Components/TabletopComputerBoardComponent.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom.Prototype;
+
+namespace Content.Server.Construction.Components;
+[RegisterComponent, ComponentProtoName("TabletopComputerBoardComponent")]
```
Comment (Tayrtahn) [inline Content.Server/Construction/Components/TabletopComputerBoardComponent.cs:5]:
```suggestion
[RegisterComponent]
```
How long ago was this code written?

**Review by Tayrtahn — CHANGES_REQUESTED** [review]

**Comment by EmoGarbage404** [comment]
i'm going to be honest, i don't really see a point in duplicating a ton of computer prototypes and adding extra components to a bunch of entities to just get an extremely mild sprite swap. the actual visual of the computer frame being cut off doesn't really look good and is really just a functional modification to justify not having any more sprites. I'd much rather we just pinpoint what computers specifically could use tabletop or mobile variants (laptops?) and then make custom sprites specifically for them.

as it stands, this is just a kind of bland modification of existing computers.

**Comment by PJB3005** [comment]
I've always though these sprites looked terrible on any server that does this concept. And this PR is no exception.

**Comment by metalgearsloth** [comment]
Okay few things:
1. The sprites don't look good.
2. As bhijn pointed out you copied frontier code and then shuffled it slightly. I do not want to catch you doing this again.
3. The code isn't even good anyway as it hardcodes functionality in.

Due to the above I am opting to close this PR.

**Pull request closed without merging**